library/std/src/sys/unsupported/rwlock.rs

```diff
@@ -56,12 +52,12 @@ impl RWLock {
 
     #[inline]
     pub unsafe fn read_unlock(&self) {
-        *self.mode.get() -= 1;
+        self.mode.set(self.mode.get() - 1);
     }
 
     #[inline]
     pub unsafe fn write_unlock(&self) {
-        *self.mode.get() += 1;
+        self.mode.set(0);
```

Comment:
This doesn't look like the same behavior.

Reply:
`write_unlock()` is only called if it was write-locked. A write-locked rwlock has a `mode` of `-1`. So that `+= 1` always results in `0`. Since `Cell` doesn't expose a `+=`, I figured this is cleaner than `self.mode.update(|n| n + 1)` or `self.mode.set(self.mode.get() + 1)`.

Reply:
Ok, can you add an assert for that then?
```suggestion
        assert_eq!(self.mode.replace(0), -1);
```

Reply:
Good idea. Thanks :)

Reply:
Maybe a `debug_assert_eq` would be enough? That's what the unix rwlocks code uses.